### PR TITLE
Test refactor

### PR DESCRIPTION
**Pull request by quantshah**
This PR will refactor the tests with a port to pytest, allowing for a cleaner and more readable test suite.

### CODE REVIEW

**Comment by josh146** [comment]
Hey @quantshah, before working too much on this, maybe we should merge in #44 so that they don't go out of sync!

**Comment by quantshah** [comment]
Yes, I will do the review now and then continue this.

**Comment by quantshah** [comment]
This is ready for review and merge. I left two tests which still need to be refactored -`test_backend_options` and `test_simple_circuits`. I would like to work on them in a different PR. They are a bit convoluted and we need to implement some way to get the IBM token working so that tests can run on the real devices too as we discussed. 

But I want to get this merged first. Feel free to suggest changes @josh146

File: tests/test_basis_state.py [hunk 1]
```diff
@@ -14,105 +14,92 @@
...
+
+elif "IBMQX_TOKEN" in os.environ and os.environ["IBMQX_TOKEN"] is not None:
+    IBMQX_TOKEN = os.environ["IBMQX_TOKEN"]
+
+devices = [BasicAerDevice(wires=num_wires), AerDevice(wires=num_wires)]
+
+if IBMQX_TOKEN is not None:
+    devices.append(
+        IbmQQiskitDevice(wires=num_wires, num_runs=8 * 1024, ibmqx_token=ibmqx_token)
+    )
```
[the remark this answers is not in the document]
Comment (josh146) [reply tests/test_basis_state.py]: The above also allows for the different backends of the providers to be tested (currently, only the default Aer `qasm_simulator` backend is being tested).

File: tests/test_expectations.py [hunk 2]
```diff
@@ -20,195 +20,239 @@
 from defaults import pennylane as qml, BaseTest
 from pennylane_qiskit import BasicAerDevice, AerDevice, IBMQDevice
 import unittest
+import pytest
+import os
 
-log.getLogger('defaults')
+log.getLogger("defaults")
```
Comment (josh146) [inline tests/test_expectations.py]: We should be able to remove the logger calls in the tests now

File: tests/test_expectations.py [hunk 3]
```diff
@@ -20,195 +20,239 @@
 from defaults import pennylane as qml, BaseTest
```
Comment (josh146) [inline tests/test_expectations.py]: Best to replace this with `import pennylane as qml` here

File: tests/test_expectations.py [hunk 4]
```diff
@@ -20,195 +20,239 @@
...
+           AerDevice(wires=num_wires, shots=shots)]
 
-    def setUp(self):
-        super().setUp()
+if IBMQX_TOKEN is not None:
+    devices.append(
+        IbmQQiskitDevice(wires=num_wires,
+                         num_runs=ibmq_shots,
+                         ibmqx_token=ibmqx_token)
+    )
```
Comment (josh146) [inline tests/test_expectations.py]: Same comment re: previous test about testing additional backends/IBMQ token.

File: tests/test_expectations.py [hunk 5]
```diff
@@ -20,195 +20,239 @@
...
+ibmq_shots = 100
+devices = [BasicAerDevice(wires=num_wires, shots=shots),
+           AerDevice(wires=num_wires, shots=shots)]
 
-    def setUp(self):
-        super().setUp()
+if IBMQX_TOKEN is not None:
+    devices.append(
+        IbmQQiskitDevice(wires=num_wires,
+                         num_runs=ibmq_shots,
```
Comment (josh146) [inline tests/test_expectations.py]: this should be shots rather than num_runs

File: tests/test_expectations.py [hunk 6]
```diff
@@ -20,195 +20,239 @@
...
-
-            O = qml.PauliZ
-            name = 'PauliZ'
+    # below are the analytic expectation values for this circuit (trace should always be 1)
+    assert np.allclose(
+        res,
+        np.array([1, 1]),
+        atol=3 / np.sqrt(device.shots),
+        rtol=3 / np.sqrt(device.shots),
+    )
```
Comment (josh146) [inline tests/test_expectations.py]: nice!

File: tests/test_expectations.py [hunk 7]
```diff
@@ -20,195 +20,239 @@
...
+    device.apply("CNOT", wires=[0, 1], par=[])
+
+    O = qml.PauliZ
+    name = "PauliZ"
+
+    device._obs_queue = [O(wires=[0], do_queue=False), O(wires=[1], do_queue=False)]
+    res = device.pre_measure()
+
+    res = np.array([device.expval(name, [0], []), device.expval(name, [1], [])])
+    print(res)
```
Comment (josh146) [inline tests/test_expectations.py]: Don't forget to remove print statements!

File: tests/test_expectations.py [hunk 8]
```diff
@@ -20,195 +20,239 @@
...
+    device.pre_measure()
+
+    res = np.array([device.expval(name, [0], []), device.expval(name, [1], [])])
+    # below are the analytic expectation values for this circuit
+    assert np.allclose(
+        res,
+        np.array([np.sin(theta) * np.sin(phi), np.sin(phi)]),
+        atol=3 / np.sqrt(device.shots),
+        rtol=3 / np.sqrt(device.shots),
+    )
```
Comment (josh146) [inline tests/test_expectations.py]: Awesome tests :)

File: tests/test_basis_state.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (josh146) [reply tests/test_basis_state.py]: Thinking about this a bit more, you could maybe turn the above into fixtures, and move them into conftest.py. Further, you could attempt to instantiate a IBMQ device, and if it fails, mark the IBMQ device fixture to be skipped via pytest.skip.

File: tests/test_simple_circuits.py [hunk 9]
```diff
@@ -26,9 +26,198 @@
...
+    """Tests if correct error is raised when attempting multi-qubit unitaries"""
+    test_input = np.array([[1, 0, 0], [0, 1, 0], [0, 0, 1]])
+    @qml.qnode(device)
+    def circuit():
+        qml.QubitUnitary(test_input, wires=[0])
+        return qml.expval(qml.PauliZ(0))
+
+    with pytest.raises(ValueError,
+                       match="Only a single qubit unitary is supported."):
+        circuit()
```
Comment (josh146) [inline tests/test_simple_circuits.py]: Hopefully we should be able to change this once qiskit v0.12 is released :) It contains support for a two-qubit `QuantumCircuit.unitary` operation, as well as an `isometry` module for arbitrary unitary decomposition.

File: tests/test_simple_circuits.py [hunk 10]
```diff
@@ -26,9 +26,198 @@
...
+    def circuit():
+        qml.QubitUnitary(test_input, wires=[0])
+        return qml.expval(qml.PauliZ(0))
+
+    with pytest.raises(ValueError,
+                       match="Not a unitary"):
+        circuit()
+
+
+class CompareWithDefaultQubitTest(BaseTest):
```
Comment (josh146) [inline tests/test_simple_circuits.py]: Agree that this is too convoluted - I would like to remove this test, and replace it with short explicit test functions. These can likely be copied over from PennyLane core, or PennyLane-Forest. But as you mentioned, better to tackle in a later PR.

File: tests/test_simple_circuits.py [hunk 11]
```diff
@@ -26,9 +26,198 @@
...
+    """Compares the behavior of the ProjectQ plugin devices with the default qubit device.
```
Comment (josh146) [inline tests/test_simple_circuits.py]: Looks like this was originally copied from ProjectQ :laughing:

**Comment by josh146** [comment]
Closing as this has been superceded by a recent test refactor

**Pull request closed without merging**